File: Practica1/S1/src/carrerasbicicletas/CarrerasBicicletas.java
package carrerasbicicletas;

import java.util.Random;

public class CarrerasBicicletas {

    public static void main(String[] args) {        
        // Tipo de carrera = 0 (Montaña) y 1 (Carretera)
        Random rand = new Random();
        int tipo_carrera = rand.nextInt(2);
        
        // Crear factoría de carreras y bicicletas
        FactoriaCarreraYBicicleta factoria = null;
        // De montaña
        if (tipo_carrera == 0){
            System.out.println("CARRERA DE MONTAÑA.");
            factoria = new FactoriaMontaña();
        }
        // De carretera
        else if (tipo_carrera == 1){
            System.out.println("CARRERA DE CARRETERA.");
            factoria = new FactoriaCarretera();
        }
        
        if (factoria != null){
            // Inicializar carrera
            Carrera carrera = factoria.crearCarrera(factoria);
            carrera.iniciarCarrera();
        }
    }
}

File: Practica1/S1E2/lib/CarrerasBicicletas.rb
require_relative "FactoriaMontaña.rb"
require_relative "FactoriaCarretera.rb"
require_relative "CarreraCarretera.rb"
require_relative "CarreraMontaña.rb"

class CarrerasBicicletas
  def self.main
    # Tipo de carrera = 0 (Montaña) y 1 (Carretera)
    tipo_carrera = Random.rand(2)
    
    # Crear factoría de carreras y bicicletas
    if tipo_carrera == 0
      puts "CARRERA DE MONTAÑA"
      factoria = FactoriaMontaña.new
    else if tipo_carrera == 1
      puts "CARRERA DE CARRETERA"
    end
      factoria = FactoriaCarretera.new
    end
    
    # Empezar carrera
    if factoria != nil
      carrera = factoria.crearCarrera
      carrera.iniciar
    end
  end
end

# Invocar al main
CarrerasBicicletas.main